annotation/geo.go

```diff
@@ -0,0 +1,237 @@
+			Labels{"source": "Failed to parse JSON"}).Inc()
+		log.Println(err)
+		return nil
+	}
+	return metaDataFromResponse
+}
+
+// QueryAnnotationService will connect to the annotation service and
+// copy the body of a valid response to a byte slice and return it to a
+// user, returning an error if any occurs
+func QueryAnnotationService(url string) ([]byte, error) {
+	// Make the actual request
+	resp, err := http.Get(url)
+
+	// Catch http errors
+	if err != nil {
+		metrics.AnnotationErrorCount.
+			With(prometheus.Labels{"source": "Request to Annotator failed"}).Inc()
+		return nil, err
+	}
+
+	// Catch errors reported by the service
+	if resp.StatusCode != http.StatusOK {
+		metrics.AnnotationErrorCount.
+			With(prometheus.Labels{"source": "Webserver gave non-ok response"}).Inc()
+		return nil, errors.New("URL:" + url + " gave response code " + resp.Status)
+	}
+	defer resp.Body.Close()
+
+	// Copy response into a byte slice
+	return ioutil.ReadAll(resp.Body)
+}
+
+// ParseJSONMetaDataResponse takes a byte slice containing the test of
+// the JSON from the annotator service and parses it into a MetaData
+// struct, for easy manipulation. It returns a pointer to the struct on
+// success and an error if an error occurs.
+func ParseJSONMetaDataResponse(jsonBuffer []byte) (*MetaData, error) {
+	parsedJSON := &MetaData{}
+	err := json.Unmarshal(jsonBuffer, parsedJSON)
+	if err != nil {
+		return nil, err
+	}
+	return parsedJSON, nil
+}
+
+// GetBatchMetaData combines the functionality of
+// BatchQueryAnnotationService and BatchParseJSONMetaDataResponse to
+// query the annotator service and return the corresponding map of
+// ip-timestamp strings to MetaData structs, or a nil map if it
+// encounters any error and cannot get the data for any reason
+// TODO - dedup common code in GetMetaData
+func GetBatchMetaData(url string, data []RequestData) map[string]MetaData {
+	// Query the service and grab the response safely
+	annotatorResponse, err := BatchQueryAnnotationService(url, data)
+	if err != nil {
+		metrics.AnnotationErrorCount.
+			With(prometheus.Labels{"source": "Error querying annotation service"}).Inc()
+		log.Println(err)
+		return nil
+	}
+
+	// Safely parse the JSON response and pass it back to the caller
+	metaDataFromResponse, err := BatchParseJSONMetaDataResponse(annotatorResponse)
+	if err != nil {
+		metrics.AnnotationErrorCount.With(prometheus.
+			Labels{"source": "Failed to parse JSON"}).Inc()
+		log.Println(err)
+		return nil
+	}
+	return metaDataFromResponse
+}
+
+// BatchQueryAnnotationService takes a url to POST the request to and
+// a slice of RequestDatas to be sent in the body in a JSON
+// format. It will copy the response into a []byte and return it to
+// the user, returning an error if any occurs
+func BatchQueryAnnotationService(url string, data []RequestData) ([]byte, error) {
+	encodedData, err := json.Marshal(data)
+	if err != nil {
+		metrics.AnnotationErrorCount.
```

Comment:
There is only AnnotationErrorCount. We need to count how many annotation requests as well.